File: panel-go/api/domains.go
package api

import (
	"strconv"

	"weavenet/panel/config"
	"weavenet/panel/database"
	"weavenet/panel/middleware"
	"weavenet/panel/models"
	"weavenet/panel/services"
	"weavenet/panel/utils"

	"github.com/gin-gonic/gin"
)

// ListDomains 免费域名列表。
func ListDomains(c *gin.Context) {
	user := middleware.GetUser(c)
	var list []models.Domain
	database.DB.Where("user_id = ?", user.ID).Order("id desc").Find(&list)
	items := make([]gin.H, 0, len(list))
	for _, d := range list {
		items = append(items, serializeDomain(&d))
	}
	utils.OK(c, items, "ok")
}

// CreateDomain 申请免费域名。
func CreateDomain(c *gin.Context) {
	var p struct {
		Subdomain string `json:"subdomain"`
		TunnelID  *uint  `json:"tunnel_id"`
	}
	if err := c.ShouldBindJSON(&p); err != nil {
		utils.Fail(c, utils.Biz(400, 0, "参数错误"))
		return
	}
	if len(p.Subdomain) < 3 || len(p.Subdomain) > 128 {
		utils.Fail(c, utils.Biz(400, 0, "子域名长度需在 3-128 位之间"))
		return
	}
	user := middleware.GetUser(c)
	db := database.DB
	var plan models.Plan
	if err := db.First(&plan, user.PlanID).Error; err != nil {
		utils.Fail(c, utils.Biz(400, 2001, "套餐配置异常，请联系管理员"))
		return
	}
	if err := services.CheckDomainLimit(db, user, &plan); err != nil {
		utils.Fail(c, utils.AsBiz(err))
		return
	}
	if !services.CheckSubdomainAvailable(db, p.Subdomain, 0) {
		utils.Fail(c, utils.Conflict(3002, "子域名已被占用，请更换"))
		return
	}
	fullDomain := p.Subdomain + "." + config.C.DomainSuffix
	d := models.Domain{
		UserID: user.ID, TunnelID: p.TunnelID, Subdomain: p.Subdomain,
		FullDomain: fullDomain, Status: "active",
	}
	if p.TunnelID != nil {
		var t models.Tunnel
		if err := db.Where("id = ? AND user_id = ?", *p.TunnelID, user.ID).First(&t).Error; err != nil {
			utils.Fail(c, utils.NotFound("隧道不存在"))
			return
		}
		if t.Type != "http" && t.Type != "https" {
			utils.Fail(c, utils.Biz(400, 0, "仅 HTTP/HTTPS 隧道可绑定免费域名"))
			return
		}
		sub := p.Subdomain
		t.Subdomain = &sub
		db.Save(&t)
	}
	if err := db.Create(&d).Error; err != nil {
		utils.Fail(c, utils.AsBiz(err))
		return
	}
	utils.OK(c, gin.H{"id": d.ID, "subdomain": d.Subdomain, "full_domain": d.FullDomain}, "免费域名申请成功")
}

// ReleaseDomain 释放免费域名。
func ReleaseDomain(c *gin.Context) {
	user := middleware.GetUser(c)
	id, err := strconv.Atoi(c.Param("domain_id"))
	if err != nil {
		utils.Fail(c, utils.NotFound("域名不存在"))
		return
	}
	db := database.DB
	var d models.Domain
	if err := db.Where("id = ? AND user_id = ?", id, user.ID).First(&d).Error; err != nil {
		utils.Fail(c, utils.NotFound("域名不存在"))
		return
	}
	d.Status = "released"
	if d.TunnelID != nil {
		var t models.Tunnel
		if err := db.First(&t, *d.TunnelID).Error; err == nil && t.Subdomain != nil && *t.Subdomain == d.Subdomain {
			t.Subdomain = nil
			db.Save(&t)
		}
	}
	db.Save(&d)
	utils.NoContent(c)
}

func serializeDomain(d *models.Domain) gin.H {
	return gin.H{
		"id": d.ID, "user_id": d.UserID, "tunnel_id": d.TunnelID, "subdomain": d.Subdomain,
		"full_domain": d.FullDomain, "status": d.Status,
		"created_at": utils.TimeFmtV(d.CreatedAt),
	}
}

File: panel/app/api/domains.py
"""免费域名 API：申请/释放。"""
from __future__ import annotations

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from ..core.config import get_settings
from ..core.database import get_db
from ..core.deps import get_current_user
from ..core.errors import BizError, ConflictError, NotFoundError, success
from ..models import Domain, Tunnel, User
from ..schemas.schemas import DomainCreate
from ..services.frp_config import check_domain_limit, check_subdomain_available

router = APIRouter(prefix="/api/domains", tags=["免费域名"])
settings = get_settings()


@router.get("")
def list_domains(user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """免费域名列表。"""
    domains = (
        db.query(Domain)
        .filter(Domain.user_id == user.id)
        .order_by(Domain.id.desc())
        .all()
    )
    return success(
        [
            {
                "id": d.id,
                "user_id": d.user_id,
                "tunnel_id": d.tunnel_id,
                "subdomain": d.subdomain,
                "full_domain": d.full_domain,
                "status": d.status,
                "created_at": d.created_at.isoformat() if d.created_at else None,
            }
            for d in domains
        ]
    )


@router.post("")
def create_domain(
    payload: DomainCreate,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """申请免费域名。"""
    db.refresh(user)
    plan = user.plan
    if plan is None:
        raise BizError(400, 2001, "套餐配置异常，请联系管理员")
    check_domain_limit(user, plan)

    if not check_subdomain_available(db, payload.subdomain):
        raise ConflictError(3002, "子域名已被占用，请更换")

    domain = Domain(
        user_id=user.id,
        tunnel_id=payload.tunnel_id,
        subdomain=payload.subdomain,
        full_domain=f"{payload.subdomain}.{settings.domain_suffix}",
        status="active",
    )
    # 校验域名路由与隧道绑定（若指定隧道，隧道须属于该用户且为 http/https 类型）
    if payload.tunnel_id:
        tunnel = (
            db.query(Tunnel)
            .filter(Tunnel.id == payload.tunnel_id, Tunnel.user_id == user.id)
            .first()
        )
        if tunnel is None:
            raise NotFoundError(message="隧道不存在")
        if tunnel.type not in ("http", "https"):
            raise BizError(400, 0, "仅 HTTP/HTTPS 隧道可绑定免费域名")
        tunnel.subdomain = payload.subdomain
    db.add(domain)
    db.commit()
    db.refresh(domain)
    return success(
        {
            "id": domain.id,
            "subdomain": domain.subdomain,
            "full_domain": domain.full_domain,
        },
        "免费域名申请成功",
        201,
    )


@router.delete("/{domain_id}", status_code=204)
def release_domain(
    domain_id: int,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """释放免费域名。"""
    domain = (
        db.query(Domain)
        .filter(Domain.id == domain_id, Domain.user_id == user.id)
        .first()
    )
    if domain is None:
        raise NotFoundError(message="域名不存在")
    domain.status = "released"
    # 解除隧道子域名绑定
    if domain.tunnel_id:
        tunnel = db.get(Tunnel, domain.tunnel_id)
        if tunnel and tunnel.subdomain == domain.subdomain:
            tunnel.subdomain = None
    db.commit()
